File: server/src/main/scala/kpn/api/common/changes/details/RouteChange.scala
package kpn.api.common.changes.details

import kpn.api.base.WithStringId
import kpn.api.common.ChangeType
import kpn.api.common.Fact
import kpn.api.common.RouteLocationAnalysis
import kpn.api.common.common.Ref
import kpn.api.common.common.ReferencedElements
import kpn.api.common.diff.RouteData
import kpn.api.common.diff.route.RouteDiff
import kpn.api.common.route.RouteNodeChange
import kpn.api.custom.Subset

/*
  Describes the changes made to a given route in a given changeset.
 */
case class RouteChange(
  _id: String,
  key: ChangeKey,
  changeType: ChangeType,
  name: String,
  locationAnalysis: RouteLocationAnalysis,
  addedToNetwork: Seq[Ref],
  removedFromNetwork: Seq[Ref],
  before: Option[RouteData],
  after: Option[RouteData],
  diffs: RouteDiff,
  nodeChanges: Seq[RouteNodeChange],
  facts: Seq[Fact],
  // following values are filled in by RouteChangeAnalyzer.analyzed
  happy: Boolean = false,
  investigate: Boolean = false,
  impact: Boolean = false,
  locationHappy: Boolean = false,
  locationInvestigate: Boolean = false,
  locationImpact: Boolean = false
) extends WithStringId {

  def id: Long = key.elementId

  def subsets: Seq[Subset] = {
    (before.toSeq.flatMap(_.subsets) ++ after.toSeq.flatMap(_.subsets)).distinct.sorted
  }

  def toRef: Ref = Ref(id, name)

  def isEmpty: Boolean = addedToNetwork.isEmpty &&
    removedFromNetwork.isEmpty &&
    diffs.isEmpty &&
    facts.isEmpty

  def referencedElements: ReferencedElements = {
    ReferencedElements(nodeIds = nodeIdsReferencedIn(before) ++ nodeIdsReferencedIn(after), Set(id))
  }

  private def nodeIdsReferencedIn(routeData: Option[RouteData]): Set[Long] = {
    routeData.toSeq.flatMap(_.networkNodes.map(_.nodeId)).toSet
  }
}

File: backend/api/src/main/java/kpn/api/common/changes/details/RouteChange.java
package kpn.api.common.changes.details;

import kpn.api.common.ChangeType;
import kpn.api.common.Fact;
import kpn.api.common.RouteLocationAnalysis;
import kpn.api.common.common.Ref;
import kpn.api.common.common.ReferencedElements;
import kpn.api.common.diff.RouteData;
import kpn.api.common.route.RouteNode;
import kpn.api.common.route.RouteNodeChange;
import kpn.api.custom.Subset;
import kpn.core.doc.WithStringId;

import java.util.Optional;
import java.util.stream.Stream;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import lombok.Builder;
import lombok.NonNull;

/*
  Describes the changes made to a given route in a given changeset.
*/
@Builder
public record RouteChange(
  @NonNull String _id,
  @NonNull ChangeKey key,
  @NonNull ChangeType changeType,
  @NonNull String name,
  @NonNull RouteLocationAnalysis locationAnalysis,
  @NonNull ImmutableList<Ref> addedToNetwork,
  @NonNull ImmutableList<Ref> removedFromNetwork,
  @NonNull Optional<RouteData> before,
  @NonNull Optional<RouteData> after,
  @NonNull ImmutableList<RouteNodeChange> nodeChanges,
  @NonNull ImmutableList<Fact> facts,
  // following values are filled in by RouteChangeAnalyzer.analyzed
  @NonNull Boolean happy,
  @NonNull Boolean investigate,
  @NonNull Boolean impact,
  @NonNull Boolean locationHappy,
  @NonNull Boolean locationInvestigate,
  @NonNull Boolean locationImpact
) implements WithStringId {

  public Long id() {
    return key.elementId();
  }

  public Ref toRef() {
    return new Ref(id(), name);
  }

  public boolean isEmpty() {
    return addedToNetwork.isEmpty() &&
      removedFromNetwork.isEmpty() &&
      facts.isEmpty();
  }

  public ImmutableList<Subset> subsets() {
    return routeDataStream()
      .flatMap(data -> data.subsets().stream())
      .distinct()
      .sorted()
      .collect(ImmutableList.toImmutableList());
  }

  public ReferencedElements referencedElements() {
    ImmutableSet<Long> routeIds = ImmutableSet.of(id());
    ImmutableSet<Long> nodeIds = routeDataStream()
      .flatMap(data -> data.networkNodes().stream())
      .map(RouteNode::nodeId)
      .collect(ImmutableSet.toImmutableSet());
    return new ReferencedElements(nodeIds, routeIds);
  }

  private Stream<RouteData> routeDataStream() {
    return Stream.concat(before.stream(), after.stream());
  }
}
